### DISCUSSION TITLE
How to implement parallel computation of beamnrc and dosxyznrc in Windows?

### DISCUSSION THREAD

**Discussion by CQUswl** [nrc-cnrc/EGSnrc#1060] [General]
The GUI doesn't enable parallel computing in Windows, is there another way to do it?

**Comment by rtownson** [comment]
Yes, using the git bash terminal on windows (installed when you install git) you can use the egs-parallel tool. e.g.
```
egs-parallel --batch cpu -n4 -v -c 'egs_app -i slab'
```
This line is telling it to run on the 4 local CPU cores, the command inside the single quotes. Just type `egs-parallel` and it will print the usage.

**Reply by CQUswl** [reply]
Thank you for your answer! If I want to run a BEAMnrc program called "BEAM_EX10MeVe" in parallel, how should I enter it?

**Reply by rtownson** [reply]
Ah on windows it isn't in the PATH by default. Instead you can use:
```
$HEN_HOUSE/scripts/bin/egs-parallel --batch cpu -n4 -v -c 'egs_app -i slab'
```
Note if you are re-running a job, you might want to also pass the --force parameter.

**Comment by CQUswl** [comment]
> Ah on windows it isn't in the PATH by default. Instead you can use:
> 
> ```
> $HEN_HOUSE/scripts/bin/egs-parallel --batch cpu -n4 -v -c 'egs_app -i slab'
> ```
> 
> Note if you are re-running a job, you might want to also pass the --force parameter.

Special thanks to you, I successfully ran BEAMnrc using this method and got results. But I am having trouble running DOSXYZnrc in parallel. I don't know if DOSXYZnrc needs to merge the result files after the run is completed like BEAMnrc. When I run DOSXYZnrc with the same input file, I don't get results from the parallel run, all the dose uncertainty is 99%, but I get normal results from the normal run, so I don't know what's wrong. When I try to use the "combine parallel" option to try to combine the parallel results, it shows the error as shown below, I don't know what's wrong, I hope you can help me.
![image](https://github.com/nrc-cnrc/EGSnrc/assets/114200268/73fcf2a1-b8c4-4d6e-9565-aef2730a4e61)

In addition, I found that the dose uncertainty in the file "inputfile_wn.egslst" is not 99%, as shown below:
![image](https://github.com/nrc-cnrc/EGSnrc/assets/114200268/7d5140b1-c7e3-4d26-aa4d-cc7b30713a6f)

The following is the result of "inputfile.egslst":
![image](https://github.com/nrc-cnrc/EGSnrc/assets/114200268/c91c650d-57bd-47bd-be91-d45ceb39bad7)

The input file:
xha                                                                              #!GUI1.0
1
H2O521ICRU
0.521, 0.01, 0, 0
-1, -1, -1, 0
-4
0.2, 40
-4
0.2, 40
0
0.2, 60
0, 0, 0, 0, 0, 0, 0, 0
0, 0, 0, 0, 0, 0, 0, 0
-6, 6, -6, 6, -6, 6, 2, 0
0, 0, 0, 0, 0, 0, 0, 0
2, 2, 0, 0, 0, 180, 0, 0, 0, 1, 3, 100, 100, 10
2, 0, 0, 0, 0, 0, 0, 0
C:/EGSnrc/egs_home/BEAM_xha/xha.egsphsp1
10000000, 0, 500, 33, 97, 100, 0, 0, 1, 1, 3, -1, 0, 0, 10, 1, 0
 #########################
 :Start MC Transport Parameter:
 
 Global ECUT= 0.521
 Global PCUT= 0.01
 Global SMAX= 5
 ESTEPE= 0.25
 XIMAX= 0.5
 Boundary crossing algorithm= PRESTA-I
 Skin depth for BCA= 0
 Electron-step algorithm= PRESTA-II
 Spin effects= On
 Brems angular sampling= Simple
 Brems cross sections= BH
 Bound Compton scattering= Off
 Compton cross sections= default
 Pair angular sampling= Simple
 Pair cross sections= BH
 Photoelectron angular sampling= Off
 Rayleigh scattering= Off
 Atomic relaxations= Off
 Electron impact ionization= Off
 Photon cross sections= xcom
 Photon cross-sections output= Off
 
 :Stop MC Transport Parameter:
 #########################

**Reply by rtownson** [reply]
Are any of the _wn.egsdat files missing after the parallel run? Maybe it is failing to add them together because they are either missing or invalid values (e.g. `nan`). Check each of the _wn.egslst files to see that the results look valid. If it seems valid for some runs and not others, maybe you just need to run more histories, or maybe the source particles are missing the phantom.

**Comment by CQUswl** [comment]
> Are any of the _wn.egsdat files missing after the parallel run? Maybe it is failing to add them together because they are either missing or invalid values (e.g. `nan`). Check each of the _wn.egslst files to see that the results look valid. If it seems valid for some runs and not others, maybe you just need to run more histories, or maybe the source particles are missing the phantom.

At the end of the parallel run I didn't find the _wn.egsdat file, only the _wn.egslst, _wn.errors, _wn.pardose and _wn.egslog files. This is shown in the image below:
![image](https://github.com/nrc-cnrc/EGSnrc/assets/114200268/e9a49f85-4b00-4014-9d4a-48d061230514)
The _wn.egslog file is in the folder shown below:
![image](https://github.com/nrc-cnrc/EGSnrc/assets/114200268/1be560e3-938f-4992-8377-f1de5d231708)

**Reply by rtownson** [reply]
Did you solve the issue? I would try running the job for combining results on the command line instead of using the GUI, just in case that is an issue.

**Reply by CQUswl** [reply]
The problem wasn't solved, so I tried running it in parallel on a Linux system. I found that the above problem does not exist in linux, so I wondered if it was a bug under Windows.
